firedrake/preconditioners/pmg.py

```diff
@@ -482,400 +495,583 @@ def coarsen_bc_value(self, bc, cV):
-                      flop_count=flop_count)
 
-
-def tensor_product_space_query(V):
+@lru_cache(maxsize=10)
+def expand_element(ele):
+    """
+    Expand a FiniteElement as an EnrichedElement of TensorProductElements, discarding modifiers.
     """
-    Checks whether the custom transfer kernels support the FunctionSpace V.
+    if ele.cell() == ufl.quadrilateral:
+        quadrilateral_tpc = ufl.TensorProductCell(ufl.interval, ufl.interval)
+        return expand_element(ele.reconstruct(cell=quadrilateral_tpc))
+    elif ele.cell() == ufl.hexahedron:
+        hexahedron_tpc = ufl.TensorProductCell(ufl.quadrilateral, ufl.interval)
+        return expand_element(ele.reconstruct(cell=hexahedron_tpc))
+    elif isinstance(ele, (ufl.TensorElement, ufl.VectorElement)):
+        return expand_element(ele._sub_element)
+    elif isinstance(ele, (ufl.HDivElement, ufl.HCurlElement, ufl.BrokenElement, ufl.RestrictedElement)):
+        return expand_element(ele._element)
+    elif isinstance(ele, ufl.WithMapping):
+        return expand_element(ele.wrapee)
+    elif isinstance(ele, ufl.MixedElement):
+        return ufl.MixedElement(*[expand_element(e) for e in ele.sub_elements()])
+    elif isinstance(ele, ufl.EnrichedElement):
+        terms = []
+        for e in ele._elements:
+            ee = expand_element(e)
+            if isinstance(ee, ufl.EnrichedElement):
+                terms.extend(ee._elements)
+            else:
+                terms.append(ee)
+        return ufl.EnrichedElement(*terms)
+    elif isinstance(ele, ufl.TensorProductElement):
+        factors = [expand_element(e) for e in ele.sub_elements()]
+        terms = [tuple()]
+        for e in factors:
+            new_terms = []
+            for f in e._elements if isinstance(e, ufl.EnrichedElement) else [e]:
+                f_factors = f.sub_elements() if isinstance(f, ufl.TensorProductElement) else (f,)
+                new_terms.extend([t_factors + f_factors for t_factors in terms])
+            terms = new_terms
+        if len(terms) == 1:
+            return ufl.TensorProductElement(*terms[0])
+        else:
+            return ufl.EnrichedElement(*[ufl.TensorProductElement(*k) for k in terms])
+    else:
+        return ele
 
-    V must be either Q(N) or DQ(N) (same N along every direction),
-    RTCF(N), RTCE(N), NCF(N), or NCE(N) on quads or hexes.
 
-    :arg V: FunctionSpace
-    :returns: 4-tuple of (use_tensorproduct, degree, topological_dimension, sub_families, variant)
+def get_line_elements(V):
+    from FIAT.reference_element import LINE
+    from tsfc.finatinterface import create_element
+    ele = V.ufl_element()
+    if isinstance(ele, ufl.MixedElement) and not isinstance(ele, (ufl.TensorElement, ufl.VectorElement)):
+        raise ValueError("MixedElements are not decomposed into tensor products")
+    rvs = ele.reference_value_size()
+    ele = expand_element(ele)
+    if isinstance(ele, ufl.EnrichedElement):
+        ele = ele._elements[0]
```

Comment:
This seems dangerous. Do you want to assert that you only have a singleton enriched element with:
```
ele, = ele._elements
```
?